store/postgres/src/relational_queries.rs

```diff
@@ -636,8 +636,21 @@ struct PrefixComparison<'a> {
 }
 
 impl<'a> PrefixComparison<'a> {
+    /// Returns the function that prefixes this column's name.
+    /// The appropriate Postgres function is `substring`, if the column's type is a byte array,
+    /// or `left` if the type is a string.
+    fn prefixing_function(column: &Column) -> &str {
+        if column.is_binary_string() {
+            "substring"
+        } else {
+            "left"
+        }
+    }
+
     fn push_column_prefix(column: &Column, mut out: AstPass<Pg>) -> QueryResult<()> {
-        out.push_sql("left(");
+        // If not a binary string (bytea), then we're prefixing text.
+        out.push_sql(Self::prefixing_function(column));
+        out.push_sql("(");
```

Comment:
This won't work; `select substring('\xdeadbeef'::bytea, 3)` returns the byte string starting at position 3. To get the equivalent of `left(s, n)` you need to use `substring(b, 1, n)`

You also introduced a constant `BYTE_ARRAY_PREFIX_SIZE` so that the equivalent of `left(s, STRING_ARRAY_PREFIX_SIZE)` should be `substring(b, 1, BYTEA_ARRAY_PREFIX_SIZE)`.